pkg/template/template.go

```diff
@@ -19,37 +20,508 @@ package template
 
 import (
 	"bytes"
+	"reflect"
+	"strings"
 	"text/template"
 
 	"github.com/Masterminds/sprig"
 	"github.com/ghodss/yaml"
+	"github.com/openebs/maya/pkg/util"
 )
 
-// toYaml takes an yaml doc in form of a string, un-marshals it to a generic
-// go struct, and marshalls back to a yaml doc. All this redundant logic
-// to get yaml in proper indentation.
+// empty returns true if the given value has the zero value for its type.
 //
-// NOTE:
-//  It will always return a string, even on error.
-//  This is designed to be called from a template.
+// This function is taken as-is from https://github.com/Masterminds/sprig
+func empty(given interface{}) bool {
+	g := reflect.ValueOf(given)
+	if !g.IsValid() {
+		return true
+	}
+
+	// Basically adapted from text/template.isTrue
+	switch g.Kind() {
+	default:
+		return g.IsNil()
+	case reflect.Array, reflect.Slice, reflect.Map, reflect.String:
+		return g.Len() == 0
+	case reflect.Bool:
+		return g.Bool() == false
+	case reflect.Complex64, reflect.Complex128:
+		return g.Complex() == 0
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return g.Int() == 0
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr:
+		return g.Uint() == 0
+	case reflect.Float32, reflect.Float64:
+		return g.Float() == 0
+	case reflect.Struct:
+		return false
+	}
+	return true
+}
+
+// NotFoundError represents an error due to a missing object
+type NotFoundError struct {
+	err string
+}
+
+func (e *NotFoundError) Error() string {
+	return e.err
+}
+
+// VerifyError represents an error due to a failure in verification
+type VerifyError struct {
+	err string
+}
+
+func (e *VerifyError) Error() string {
+	return e.err
+}
+
+// isLen returns true if the expected value matches the given object's
+// length
 //
-// NOTE: Borrowed from a similar function in helm
-//  https://github.com/kubernetes/helm/blob/master/pkg/chartutil/files.go
-func toYaml(yml string) string {
-	var m interface{}
-	err := yaml.Unmarshal([]byte(yml), &m)
+// This function is intended to be used as a go template function.
+//
+// Example:
+// {{- "abc def" | splitList " " | isLen 2 | not | verifyErr "count is not two" | noop -}}
+func isLen(expected int, given interface{}) bool {
```

Comment:
can this method be subsumed by `func empty(given interface{}) bool`? So that we have a generic method. Just a though.